pkg/operator/csidriveroperator/olmremovalcontroller.go

```diff
@@ -0,0 +1,409 @@
+	olmOptions     *csioperatorclient.OLMOptions
+	dynamicClient  dynamic.Interface
+	kubeClient     kubernetes.Interface
+	eventRecorder  events.Recorder
+	factory        *factory.Factory
+
+	olmOperatorNamespace string
+	olmOperatorCSVName   string
+}
+
+const (
+	olmOperatorRemovalControllerName = "OLMOperatorRemoval"
+
+	// Annotation used to store OLM-based operator namespace in CSO's CR
+	olmOperatorNamespaceAnnotation = ".olm-removal.storage.openshift.io/namespace"
+	// Annotation used to store OLM-based operator CSV name in CSO's CR
+	olmOperatorCSVAnnotation = ".olm-removal.storage.openshift.io/csvName"
+
+	olmSource          = "redhat-operators"
+	olmSourceNamespace = "openshift-marketplace"
+
+	oldCRName = "cluster"
+
+	// Interval used to check if an deleted objects was really removed from
+	// API server. OLMOperatorRemovalController does not have informer on all
+	// objects it needs to watch / remove (namely OLM CRDs and all Deployments
+	// on the system, it would be too noisy).
+	waitInterval = 5 * time.Second
+)
+
+func NewOLMOperatorRemovalController(
+	csiOperatorConfig csioperatorclient.CSIOperatorConfig,
+	clients *csoclients.Clients,
+	eventRecorder events.Recorder,
+	resyncInterval time.Duration,
+) *OLMOperatorRemovalController {
+
+	if csiOperatorConfig.OLMOptions == nil {
+		return nil
+	}
+
+	f := factory.New()
+	f = f.ResyncEvery(resyncInterval)
+	f = f.WithSyncDegradedOnError(clients.OperatorClient)
+	// Necessary to do initial Sync after the controller starts.
+	f = f.WithPostStartHooks(initalSync)
+	// Add informers to the factory now, but the actual event handlers
+	// are added later in CSIDriverOperatorCRController.Run(),
+	// when we're 100% sure the controller is going to start (because it
+	// depends on the platform).
+	// If we added the event handlers now, all events would pile up in the
+	// controller queue, without anything reading it.
+	// Do *not* watch all Deployments in the cluster - that would be too noisy.
+	f = f.WithInformers(clients.OperatorClient.Informer())
+
+	c := &OLMOperatorRemovalController{
+		name:           csiOperatorConfig.ConditionPrefix,
+		operatorClient: clients.OperatorClient,
+		olmOptions:     csiOperatorConfig.OLMOptions,
+		dynamicClient:  clients.DynamicClient,
+		kubeClient:     clients.KubeClient,
+		eventRecorder:  eventRecorder.WithComponentSuffix(csiOperatorConfig.ConditionPrefix),
+		factory:        f,
+	}
+	return c
+}
+
+func (c *OLMOperatorRemovalController) Sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	klog.V(4).Infof("OLMOperatorRemovalController.Sync started")
+	defer klog.V(4).Infof("OLMOperatorRemovalController.Sync finished")
+
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorapi.Managed {
+		return nil
+	}
+
+	// 1. Find subscription + namespace
+	subNamespace, subName, csvName, found, err := c.findSubscription(ctx)
+	if err != nil {
+		return err
+	}
```

Comment:
Is there a way to short-circuit here in case everything has been deleted already?